cpp/src/trajectory/compute_distance_and_speed.cu

```diff
@@ -0,0 +1,191 @@
+    compute_distance_and_speed_kernel<Element, Timestamp>
+        <<<grid.num_blocks, block_size, 0, stream>>>(
+            *cudf::column_device_view::create(x, stream),
+            *cudf::column_device_view::create(y, stream),
+            *cudf::column_device_view::create(timestamp, stream),
+            *cudf::column_device_view::create(offset, stream),
+            *cudf::mutable_column_device_view::create(*cols.at(0), stream),
+            *cudf::mutable_column_device_view::create(*cols.at(1), stream));
+
+    // check for errors
+    CHECK_CUDA(stream);
+
+    return std::make_unique<cudf::experimental::table>(std::move(cols));
+  }
+
+  template <typename Timestamp>
+  std::enable_if_t<not cudf::is_timestamp<Timestamp>(),
+                   std::unique_ptr<cudf::experimental::table>>
+  operator()(cudf::column_view const& x, cudf::column_view const& y,
+             cudf::column_view const& timestamp,
+             cudf::column_view const& offset,
+             rmm::mr::device_memory_resource* mr, cudaStream_t stream) {
+    CUDF_FAIL("Timestamp must be a timestamp type");
+  }
+};
+
+struct dispatch_element {
+  template <typename Element>
+  std::enable_if_t<std::is_floating_point<Element>::value,
+                   std::unique_ptr<cudf::experimental::table>>
+  operator()(cudf::column_view const& x, cudf::column_view const& y,
+             cudf::column_view const& timestamp,
+             cudf::column_view const& offset,
+             rmm::mr::device_memory_resource* mr, cudaStream_t stream) {
+    return cudf::experimental::type_dispatcher(
+        timestamp.type(), dispatch_timestamp<Element>{}, x, y, timestamp,
+        offset, mr, stream);
+  }
+
+  template <typename Element>
+  std::enable_if_t<not std::is_floating_point<Element>::value,
+                   std::unique_ptr<cudf::experimental::table>>
+  operator()(cudf::column_view const& x, cudf::column_view const& y,
+             cudf::column_view const& timestamp,
+             cudf::column_view const& offset,
+             rmm::mr::device_memory_resource* mr, cudaStream_t stream) {
+    CUDF_FAIL("X and Y must be floating point types");
+  }
+};
+
+}  // namespace
+
+namespace detail {
+std::unique_ptr<cudf::experimental::table> compute_distance_and_speed(
+    cudf::column_view const& x, cudf::column_view const& y,
+    cudf::column_view const& timestamp, cudf::column_view const& offset,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream) {
+  return cudf::experimental::type_dispatcher(x.type(), dispatch_element{}, x, y,
+                                             timestamp, offset, mr, stream);
+}
+}  // namespace detail
+
+std::unique_ptr<cudf::experimental::table> compute_distance_and_speed(
+    cudf::column_view const& x, cudf::column_view const& y,
+    cudf::column_view const& timestamp, cudf::column_view const& offset,
+    rmm::mr::device_memory_resource* mr) {
+  CUSPATIAL_EXPECTS(!(x.is_empty() || y.is_empty() || offset.is_empty() ||
```

Comment:
Generally in cuDF for empty inputs we return empty outputs, and don't throw an exception. This is to support things like Dask that sometimes generates empty tasks.  I am not sure, but I think we should probably follow the same process here.